src/generic/fetch.ts

```diff
@@ -0,0 +1,57 @@
+/**
+ * Contains utility methods using 'fetch'.
+ */
+
+import {ApiError, ErrorCode} from '../core/api_error';
+import {BFSCallback} from '../core/file_system';
+
+/**
+ * Asynchronously download a file as a buffer or a JSON object.
+ * Note that the third function signature with a non-specialized type is
+ * invalid, but TypeScript requires it when you specialize string arguments to
+ * constants.
+ * @hidden
+ */
+export function fetchFileAsync(p: string, type: 'buffer', cb: BFSCallback<Buffer>): void;
+export function fetchFileAsync(p: string, type: 'json', cb: BFSCallback<any>): void;
+export function fetchFileAsync(p: string, type: string, cb: BFSCallback<any>): void;
+export function fetchFileAsync(p: string, type: string, cb: BFSCallback<any>): void {
+  fetch(p)
+    .then((res) => {
+      if (!res.ok) {
+        return cb(new ApiError(res.status, "fetch error."));
```

Comment:
Can you change this to use `ErrorCode.EIO`? I realize that I did this in my `xhr` code, but that should really be fixed. The error code should always be one of those libc error codes. You can include the status code in the error message, e.g.:

```typescript
new ApiError(ErrorCode.EIO, `fetch error: response returned code ${res.status}`)
```

If you could update my XHR code to do the same, I'd be very grateful. I haven't looked over that part of the code in years!